src/mmc.c

```diff
@@ -0,0 +1,272 @@
+  sstrncpy(vl.type, type, sizeof(vl.type));
+
+  plugin_dispatch_values(&vl);
+}
+
+static int mmc_read_dev_attr(const char *dev_name, const char *file_name,
+                             char *buffer, int size) {
+  FILE *fh;
+  char str[sizeof(SYS_PATH) + strlen(dev_name) + sizeof("/") +
+           strlen(file_name) + 1];
+  int length;
+
+  snprintf(str, sizeof(str), SYS_PATH "%s/%s", dev_name, file_name);
+  fh = fopen(str, "r");
+
+  if (fh == NULL) {
+    ERROR(PLUGIN_NAME "(%s): Cannot open file [%s]", dev_name, str);
+    return EXIT_FAILURE;
+  }
+
+  DEBUG(PLUGIN_NAME "(%s): try to read [%s]", dev_name, str);
+  if (fgets(buffer, size, fh) == NULL) {
+    ERROR(PLUGIN_NAME "(%s): Unable to read file [%s] (%s)", dev_name, str,
+          STRERRNO);
+    fclose(fh);
+    return EXIT_FAILURE;
+  }
+  fclose(fh);
+
+  /* Remove trailing whitespace for sysfs attr read */
+  length = strlen(buffer);
+  DEBUG(PLUGIN_NAME "(%s): Read %d characters [%s]", dev_name, length, str);
+  if (buffer > 0)
+    buffer[length - 1] = '\0';
+
+  return 0;
+}
+
+static int mmc_read_manfid(const char *dev_name, int *value) {
+  char buffer[4096];
+
+  if (mmc_read_dev_attr(dev_name, MMC_MANUFACTOR, buffer, sizeof(buffer)) ==
+      0) {
+    *value = (int)strtol(buffer, NULL, 0);
+    DEBUG(PLUGIN_NAME "(%s): [%s]=%s (%d)", dev_name, MMC_MANUFACTOR, buffer,
+          *value);
+    return 0;
+  }
+
+  WARNING(PLUGIN_NAME "(%s): Unable to read manfid", dev_name);
+  return EXIT_FAILURE;
+}
+
+static int mmc_read_oemid(const char *dev_name, int *value) {
+  char buffer[4096];
+
+  if (mmc_read_dev_attr(dev_name, MMC_OEM_ID, buffer, sizeof(buffer)) == 0) {
+    *value = (int)strtol(buffer, NULL, 0);
+    DEBUG(PLUGIN_NAME "(%s): [%s]=%s (%d)", dev_name, MMC_OEM_ID, buffer,
+          *value);
+    return 0;
+  }
+
+  WARNING(PLUGIN_NAME "(%s): Unable to read oemid", dev_name);
+  return EXIT_FAILURE;
+}
+
+enum mmc_manfid {
+  MANUFACTUR_SWISSBIT = 93, // 0x5d
+};
+
+enum mmc_oemid_swissbit {
+  OEMID_SWISSBIT_1 = 21314, // 0x5342
+};
+
+#define MMC_POWER_CYCLES "mmc_power_cycles"
+#define MMC_BLOCK_ERASES "mmc_block_erases"
+#define MMC_BAD_BLOCKS "mmc_bad_blocks"
+
+// Size of string buffer with '\0'
+#define SWISSBIT_LENGTH_SPARE_BLOCKS 3
+#define SWISSBIT_LENGTH_BLOCK_ERASES 13
+#define SWISSBIT_LENGTH_POWER_ON 9
+
+#define SWISSBIT_SSR_START_SPARE_BLOCKS 66
+#define SWISSBIT_SSR_START_BLOCK_ERASES 92
+#define SWISSBIT_SSR_START_POWER_ON 112
+
+static int mmc_read_ssr_swissbit(const char *dev_name) {
+  char buffer[4096];
+  int oemid;
+  int value;
+  int length;
+  char bad_blocks[SWISSBIT_LENGTH_SPARE_BLOCKS];
+  char block_erases[SWISSBIT_LENGTH_BLOCK_ERASES];
+  char power_on[SWISSBIT_LENGTH_POWER_ON];
+
+  if (mmc_read_oemid(dev_name, &oemid) != 0) {
+    return EXIT_FAILURE;
+  }
+
+  if (oemid != OEMID_SWISSBIT_1) {
+    INFO(PLUGIN_NAME "(%s): The oemid id %d is not suppored by this plugin",
```

Comment:
Same here. Can we make the output a little more verbose? At the end, this will be read by a human being.